File: ports/zig/src/widgets/surface.zig
//! Two primitives for the space behind a widget rather than the widget itself.
//!
//! `modal` already blanks the region it is about to draw into, but it does it
//! privately, so anything else that floats -- a custom overlay, a popover, a
//! tooltip somebody wrote themselves -- has no way to say "this region is mine
//! now". These make that sayable.

const std = @import("std");

const buffer_mod = @import("../buffer.zig");
const color_mod = @import("../color.zig");
const surface_mod = @import("../surface.zig");
const unicode = @import("../unicode.zig");

const Attrs = buffer_mod.Attrs;
const Color = color_mod.Color;
const Style = buffer_mod.Style;
const Surface = surface_mod.Surface;

pub const ClearOptions = struct {
    /// What to leave behind. Defaults to the theme's background.
    background: ?Color = null,
};

/// Reset a region to empty, so an overlay can draw over what was there.
///
/// Without this an overlay is drawn *into* whatever it lands on: the cells it
/// does not touch keep the widget underneath, and a dialog ends up with someone
/// else's table showing through the gaps between its words.
pub fn drawClear(s: Surface, options: ClearOptions) void {
    if (s.isEmpty()) return;
    const theme = s.theme;
    s.fill(.{
        .fg = theme.foreground,
        .bg = options.background orelse theme.background,
        .attrs = Attrs.none,
    });
}

pub const FillOptions = struct {
    /// The symbol to repeat. A wide one is stepped over rather than written per
    /// column, since each glyph owns a continuation cell.
    symbol: []const u8 = " ",
    fg: ?Color = null,
    bg: ?Color = null,
    attrs: ?Attrs = null,
};

/// Flood a region with one repeated symbol and style.
pub fn drawFill(s: Surface, options: FillOptions) void {
    if (s.isEmpty()) return;
    const symbol = if (options.symbol.len > 0) options.symbol else " ";
    const style = Style{ .fg = options.fg, .bg = options.bg, .attrs = options.attrs };

    const len = std.unicode.utf8ByteSequenceLength(symbol[0]) catch 1;
    const glyph: u21 = if (len <= symbol.len)
        std.unicode.utf8Decode(symbol[0..len]) catch ' '
    else
        ' ';
    const glyph_width = @max(1, unicode.stringWidth(symbol));

    // A one-cell symbol is what `fill` is for. Anything wider has to be stepped
    // over rather than written per column: each glyph owns a continuation cell,
    // and writing the next one on top of it leaves a row of half-characters.
    if (glyph_width == 1) {
        s.fillRect(0, 0, s.width(), s.height(), style, glyph);
        return;
    }
    for (0..s.height()) |y| {
        // The last glyph is dropped rather than clipped when the region does
        // not divide evenly: half a wide character is not a fill, it is damage.
        var x: usize = 0;
        while (x + glyph_width <= s.width()) : (x += glyph_width) {
            s.glyph(@intCast(x), @intCast(y), glyph, style);
        }
    }
}

File: ports/rust/src/widgets/surface.rs
//! Two primitives for the space behind a widget rather than the widget itself.
//!
//! `modal` already blanks the region it is about to draw into, but it does it
//! privately, so anything else that floats -- a custom overlay, a popover, a
//! tooltip somebody wrote themselves -- has no way to say "this region is mine
//! now". These make that sayable.

use crate::buffer::{Attrs, Style};
use crate::color::Color;
use crate::surface::Surface;
use crate::unicode::string_width;

#[derive(Clone, Copy, Debug, Default)]
pub struct ClearOptions {
    /// What to leave behind. Defaults to the theme's background.
    pub background: Option<Color>,
}

/// Reset a region to empty, so an overlay can draw over what was there.
///
/// Without this an overlay is drawn *into* whatever it lands on: the cells it
/// does not touch keep the widget underneath, and a dialog ends up with someone
/// else's table showing through the gaps between its words.
pub fn draw_clear(surface: &Surface, options: &ClearOptions) {
    if surface.is_empty() {
        return;
    }
    let theme = surface.theme.clone();
    surface.fill(&Style {
        fg: Some(theme.foreground),
        bg: Some(options.background.unwrap_or(theme.background)),
        attrs: Some(Attrs::default()),
    });
}

#[derive(Clone, Debug)]
pub struct FillOptions {
    /// The symbol to repeat. A wide one is stepped over rather than written per
    /// column, since each glyph owns a continuation cell.
    pub symbol: String,
    pub fg: Option<Color>,
    pub bg: Option<Color>,
    pub attrs: Option<Attrs>,
}

impl Default for FillOptions {
    fn default() -> FillOptions {
        FillOptions { symbol: " ".into(), fg: None, bg: None, attrs: None }
    }
}

/// Flood a region with one repeated symbol and style.
pub fn draw_fill(surface: &Surface, options: &FillOptions) {
    if surface.is_empty() {
        return;
    }
    let style = Style { fg: options.fg, bg: options.bg, attrs: options.attrs };
    let glyph = options.symbol.chars().next().unwrap_or(' ');
    let glyph_width = string_width(&options.symbol).max(1);
    // A one-cell symbol is what `fill` is for. Anything wider has to be stepped
    // over rather than written per column: each glyph owns a continuation cell,
    // and writing the next one on top of it leaves a row of half-characters.
    if glyph_width == 1 {
        surface.fill_rect(0, 0, surface.width(), surface.height(), &style, glyph as u32);
        return;
    }
    for y in 0..surface.height() {
        // The last glyph is dropped rather than clipped when the region does
        // not divide evenly: half a wide character is not a fill, it is damage.
        let mut x = 0usize;
        while x + glyph_width <= surface.width() {
            surface.glyph(x as isize, y as isize, glyph, &style);
            x += glyph_width;
        }
    }
}
